src/compiler/binder.ts

```diff
@@ -329,12 +351,273 @@ namespace ts {
+                    return bindLabeledStatement(<LabeledStatement>n);
+                default:
+                    return false;
+            }
+        }
+
+        function bindWhileStatement(n: WhileStatement): boolean {
+            const preWhileState = 
+                n.expression.kind === SyntaxKind.FalseKeyword ? Reachability.Unreachable : currentReachabilityState;
+            const postWhileState = 
+                n.expression.kind === SyntaxKind.TrueKeyword ? Reachability.Unreachable : currentReachabilityState;
+
+            // bind expressions (don't affect reachability)
+            bind(n.expression);
+            
+            currentReachabilityState = preWhileState;
+            const postWhileLabel = pushImplicitLabel();
+            bind(n.statement);
+            popImplicitLabel(postWhileLabel, postWhileState)
+            
+            return true;
+        }
+
+        function bindDoStatement(n: DoStatement): boolean {
+            const preDoState = currentReachabilityState;
+
+            const postDoLabel = pushImplicitLabel();
+            bind(n.statement);
+            var postDoState = n.expression.kind === SyntaxKind.TrueKeyword ? Reachability.Unreachable : preDoState;
+            popImplicitLabel(postDoLabel, postDoState);
+
+            // bind expressions (don't affect reachability)
+            bind(n.expression);
+
+            return true;
+        }
+        
+        function bindForStatement(n: ForStatement): boolean {
+            const preForState = currentReachabilityState;
+            const postForLabel = pushImplicitLabel();
+
+            // bind expressions (don't affect reachability)
+            bind(n.initializer);
+            bind(n.condition);
+            bind(n.incrementor);
+
+            bind(n.statement);
+
+            // for statement is considered infinite when it condition is either omitted or is true keyword
+            // - for(..;;..)
+            // - for(..;true;..)
+            const isInfiniteLoop = (!n.condition || n.condition.kind === SyntaxKind.TrueKeyword);
+            const postForState = isInfiniteLoop ? Reachability.Unreachable : preForState;
+            popImplicitLabel(postForLabel, postForState);
+
+            return true;
+        }
+
+        function bindForInOrForOfStatement(n: ForInStatement | ForOfStatement): boolean {
+            const preStatementState = currentReachabilityState;
+            const postStatementLabel = pushImplicitLabel();
+            
+            // bind expressions (don't affect reachability)
+            bind(n.initializer);
+            bind(n.expression);
+
+            bind(n.statement);
+            popImplicitLabel(postStatementLabel, preStatementState);
+
+            return true;
+        }
+
+        function bindIfStatement(n: IfStatement): boolean {
+            const ifTrueState = n.expression.kind === SyntaxKind.FalseKeyword ? Reachability.Unreachable : currentReachabilityState;

```

Comment:
This reads very strangely - it can be read as "the reachability of `if (true)` than the state within the `if` block. Maybe a quick comment would help.